src/libsyntax/ext/tt/macro_parser.rs

```diff
@@ -251,14 +251,25 @@ pub fn nameize(p_s: &ParseSess, ms: &[TokenTree], res: &[Rc<NamedMatch>])
 
 pub enum ParseResult<T> {
     Success(T),
-    /// Arm failed to match
-    Failure(syntax_pos::Span, String),
+    /// Arm failed to match. If the second parameter is `None`, it indicates an
+    /// unexpected of macro invocation. Otherwise, it indicates that no rules
+    /// expected the given token.
+    Failure(syntax_pos::Span, Option<Token>),
     /// Fatal error (malformed macro?). Abort compilation.
     Error(syntax_pos::Span, String)
 }
 
+impl<T> ParseResult<T> {
+    pub fn msg(tok: Option<Token>) -> String {
```

Comment:
Indeed, I suggested `failure_message` on IRC (with that same descriptivity goal).